src/Core/Geometry/IndexedGeometry.cpp

```diff
@@ -0,0 +1,263 @@
+    AttribArrayGeometry::operator=( std::move( other ) );
+    m_indices                    = std::move( other.m_indices );
+    notify();
+    invalidateAabb();
+    return *this;
+}
+
+void MultiIndexedGeometry::clear() {
+    AttribArrayGeometry::clear();
+    deepClear();
+    notify();
+    invalidateAabb();
+}
+
+void MultiIndexedGeometry::copy( const MultiIndexedGeometry& other ) {
+    AttribArrayGeometry::copyBaseGeometry( other );
+    deepCopy( other );
+    notify();
+    invalidateAabb();
+}
+
+/// \todo Implement MultiIndexedGeometry::checkConsistency
+void MultiIndexedGeometry::checkConsistency() const {
+#ifdef CORE_DEBUG
+#endif
+}
+
+bool
+/// \todo Implement MultiIndexedGeometry::append
+MultiIndexedGeometry::append( const MultiIndexedGeometry& other ) {
+    bool isCopyComplete = true;
+    for ( const auto& [key, value] : m_indices )
+    {
+        auto it = other.m_indices.find( key );
+        if ( !( it == other.m_indices.end() && value.second->append( *( it->second.second ) ) ) )
+            isCopyComplete = false;
+    }
+
+    notify();
+    invalidateAabb();
+    return isCopyComplete;
+}
+
+//////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////
+
+bool MultiIndexedGeometry::containsLayer( const LayerSemantic& semanticName ) const {
+    for ( const auto& [key, value] : m_indices )
+    {
+        if ( key.first.find( semanticName ) != key.first.end() ) return true;
+    }
+    return false;
+}
+
+bool MultiIndexedGeometry::containsLayer( const LayerSemanticCollection& semantics ) const {
+    for ( const auto& [key, value] : m_indices )
+    {
+        if ( key.first == semantics ) return true;
+    }
+    return false;
+}
+
+//////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////
+
+size_t MultiIndexedGeometry::countLayers( const LayerSemantic& semanticName ) const {
+    size_t c = 0;
+    for ( const auto& [key, value] : m_indices )
+    {
+        if ( key.first.find( semanticName ) != key.first.end() ) ++c;
+    }
+    return c;
+}
+
+size_t MultiIndexedGeometry::countLayers( const LayerSemanticCollection& semantics ) const {
+    size_t c = 0;
+    for ( const auto& [key, value] : m_indices )
+    {
+        if ( key.first == semantics ) ++c;
+    }
+    return c;
+}
+
+//////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////
+
+std::pair<MultiIndexedGeometry::LayerKeyType, const GeometryIndexLayerBase&>
+MultiIndexedGeometry::getFirstLayerOccurrence( const LayerSemantic& semanticName ) const {
+    for ( const auto& [key, value] : m_indices )
+    {
+        if ( key.first.find( semanticName ) != key.first.end() ) return {key, *( value.second )};
+    }
+    throw std::out_of_range( "Layer entry not found" );
```

Comment:
Hum right, I didn't get that I was introducing exceptions back.
I suggest to remove the exceptions in the PR, and to open a discussion to decide if we want to change the rules or not.

Reply:
@dlyr I tried to remove exceptions and to use `optional` instead, see a7a0ae4.
As shown below, this change overload client code. We discussed with @MathiasPaulin  and agreed that we both prefer the previous version with exceptions. 

```diff
- auto& abstractLayer = getLayerWithLock( m_mainIndexLayerKey );
- return static_cast<DefaultLayerType&>( abstractLayer ).collection();
+ auto abstractLayer = getLayerWithLock( m_mainIndexLayerKey );
+ if( abstractLayer )
+ {
+    return static_cast<DefaultLayerType&>( (*abstractLayer).get() ).collection();
+ }
```

What do you think ?

Reply:
Let's do the throw (by exception ;))

Reply:
As stated above, ...  this will need to update the "Coding convention" page of the Radium documentation (https://storm-irit.github.io/Radium-Engine/develCodingConvention.html) that says :
> Non-negociable
>No exceptions, no RTTI, no dynamic cast, no gotos.